File: halocline-stage1/src/lib/model/modelTuning.ts
export type ModelTuningInput = {
  initialHeadMeters?: number;
  regionalGradientMetersPerKilometer?: number;
  baseRechargeMillimetersPerYear?: number;
  hydraulicConductivityScale?: number;
  defaultCanalStageMeters?: number;
};

export type ResolvedModelTuning = Required<ModelTuningInput>;

export const defaultModelTuning = {
  initialHeadMeters: 0.6,
  regionalGradientMetersPerKilometer: 0.025,
  baseRechargeMillimetersPerYear: 1300,
  hydraulicConductivityScale: 1,
  defaultCanalStageMeters: 0.55,
} satisfies ResolvedModelTuning;

export function resolveModelTuning(input: ModelTuningInput = {}): ResolvedModelTuning {
  return {
    initialHeadMeters: input.initialHeadMeters ?? defaultModelTuning.initialHeadMeters,
    regionalGradientMetersPerKilometer:
      input.regionalGradientMetersPerKilometer ??
      defaultModelTuning.regionalGradientMetersPerKilometer,
    baseRechargeMillimetersPerYear:
      input.baseRechargeMillimetersPerYear ??
      defaultModelTuning.baseRechargeMillimetersPerYear,
    hydraulicConductivityScale:
      input.hydraulicConductivityScale ?? defaultModelTuning.hydraulicConductivityScale,
    defaultCanalStageMeters:
      input.defaultCanalStageMeters ?? defaultModelTuning.defaultCanalStageMeters,
  };
}

export function modelTuningIsDefault(tuning: ResolvedModelTuning): boolean {
  return (
    tuning.initialHeadMeters === defaultModelTuning.initialHeadMeters &&
    tuning.regionalGradientMetersPerKilometer ===
      defaultModelTuning.regionalGradientMetersPerKilometer &&
    tuning.baseRechargeMillimetersPerYear ===
      defaultModelTuning.baseRechargeMillimetersPerYear &&
    tuning.hydraulicConductivityScale === defaultModelTuning.hydraulicConductivityScale &&
    tuning.defaultCanalStageMeters === defaultModelTuning.defaultCanalStageMeters
  );
}

File: research/halocline-physics/src/halocline_physics/model_tuning.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class ResolvedModelTuning:
    initial_head_meters: float = 0.6
    regional_gradient_meters_per_kilometer: float = 0.025
    base_recharge_millimeters_per_year: float = 1300
    hydraulic_conductivity_scale: float = 1
    default_canal_stage_meters: float = 0.55


default_model_tuning = ResolvedModelTuning()


def resolve_model_tuning(
    tuning: ResolvedModelTuning | dict[str, float] | None = None,
) -> ResolvedModelTuning:
    if tuning is None:
        return default_model_tuning
    if isinstance(tuning, ResolvedModelTuning):
        return tuning
    return ResolvedModelTuning(
        initial_head_meters=tuning.get("initial_head_meters", tuning.get("initialHeadMeters", default_model_tuning.initial_head_meters)),
        regional_gradient_meters_per_kilometer=tuning.get(
            "regional_gradient_meters_per_kilometer",
            tuning.get("regionalGradientMetersPerKilometer", default_model_tuning.regional_gradient_meters_per_kilometer),
        ),
        base_recharge_millimeters_per_year=tuning.get(
            "base_recharge_millimeters_per_year",
            tuning.get("baseRechargeMillimetersPerYear", default_model_tuning.base_recharge_millimeters_per_year),
        ),
        hydraulic_conductivity_scale=tuning.get(
            "hydraulic_conductivity_scale",
            tuning.get("hydraulicConductivityScale", default_model_tuning.hydraulic_conductivity_scale),
        ),
        default_canal_stage_meters=tuning.get(
            "default_canal_stage_meters",
            tuning.get("defaultCanalStageMeters", default_model_tuning.default_canal_stage_meters),
        ),
    )
